### PR TITLE
chore(actions): parallelize test execution

### LINKED ISSUE #117: Validation workflow: run test steps in parallel
Test steps (test, end-to-end dev, end-to-end prod) take more than 1 minute to complete.
Update the workflow to run them in parallel to speed up the validation

### PR DESCRIPTION
**Pull request by mcollovati**
Executes different types of tests in parallel to speed up the validation.
Validation jobs are skipped if the PR only modifies files that do not affect the build or the tests (e.g. README).

Closes #117

### CODE REVIEW

File: .github/workflows/validation.yaml [hunk 1]
```diff
@@ -61,25 +76,116 @@ jobs:
         run: |
           set -x -e -o pipefail
           mvn -V -e -B -ntp verify -Dmaven.javadoc.skip=false -DtrimStackTrace=false
+      - name: Package test output files
+        if: ${{ failure() || success() }}
+        run: find . -name surefire-reports -o -name failsafe-reports -o -name error-screenshots | tar -czf tests-report-main.tgz -T -
```
Comment (Dudeplayz) [inline .github/workflows/validation.yaml:81]:
```suggestion
        run: find . -name surefire-reports -o -name failsafe-reports -o -name error-screenshots | tar -czf tests-report-main.tgz -T
```

Last `-` wanted?

File: .github/workflows/validation.yaml [hunk 2]
```diff
@@ -61,25 +76,116 @@ jobs:
...
+      - uses: actions/upload-artifact@v3
+        if: ${{ failure() || success() }}
```
Comment (Dudeplayz) [inline .github/workflows/validation.yaml:83]:
```suggestion
```

Additional check for uses required?

File: .github/workflows/validation.yaml [hunk 3]
```diff
@@ -61,25 +76,116 @@ jobs:
...
+          distribution: 'temurin'
+          cache: maven
+      - uses: browser-actions/setup-chrome@latest
+        with:
+          chrome-version: stable
+      - name: Build
+        run: |
+          set -x -e -o pipefail
+          mvn -V -e -B -ntp -DskipTests -Dmaven.javadoc.skip=false install
       - name: End to end Test (Development mode)
```
Comment (Dudeplayz) [inline .github/workflows/validation.yaml:111]:
```suggestion
      - name: End-to-end Test (Development mode)
```

File: .github/workflows/validation.yaml [hunk 4]
```diff
@@ -61,25 +76,116 @@ jobs:
...
         run: |
           set -x -e -o pipefail
           mvn -V -e -B -ntp verify -Dmaven.javadoc.skip=false -DtrimStackTrace=false -Pit-tests
+      - name: Package test output files
+        if: ${{ failure() || success() }}
+        run: find . -name surefire-reports -o -name failsafe-reports -o -name error-screenshots | tar -czf tests-report-e2e-dev.tgz -T -
```
Comment (Dudeplayz) [inline .github/workflows/validation.yaml:117]:
```suggestion
        run: find . -name surefire-reports -o -name failsafe-reports -o -name error-screenshots | tar -czf tests-report-e2e-dev.tgz -T
```

?

File: .github/workflows/validation.yaml [hunk 5]
```diff
@@ -61,25 +76,116 @@ jobs:
...
+          distribution: 'temurin'
+          cache: maven
+      - uses: browser-actions/setup-chrome@latest
+        with:
+          chrome-version: stable
+      - name: Build
+        run: |
+          set -x -e -o pipefail
+          mvn -V -e -B -ntp -DskipTests -Dmaven.javadoc.skip=false install
       - name: End to end Test (Production mode)
```
Comment (Dudeplayz) [inline .github/workflows/validation.yaml:147]:
```suggestion
      - name: End-to-end Test (Production mode)
```

File: .github/workflows/validation.yaml [hunk 6]
```diff
@@ -61,25 +76,116 @@ jobs:
...
         run: |
           set -x -e -o pipefail
           mvn -V -e -B -ntp verify -Dmaven.javadoc.skip=false -DtrimStackTrace=false -Pit-tests,production
       - name: Package test output files
         if: ${{ failure() || success() }}
-        run: find . -name surefire-reports -o -name failsafe-reports -o -name error-screenshots | tar -czf tests-report.tgz -T -
+        run: find . -name surefire-reports -o -name failsafe-reports -o -name error-screenshots | tar -czf tests-report-e2e-prod.tgz -T -
```
Comment (Dudeplayz) [inline .github/workflows/validation.yaml:153]:
```suggestion
        run: find . -name surefire-reports -o -name failsafe-reports -o -name error-screenshots | tar -czf tests-report-e2e-prod.tgz -T
```

?

File: .github/workflows/validation.yaml [hunk 7]
```diff
@@ -61,25 +76,116 @@ jobs:
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
+      - uses: actions/download-artifact@v3
+        if: ${{ always() && needs.changes.outputs.validation-required == 'true' }}
+        with:
+          name: tests-output
+      - name: extract downloaded files
```
Comment (Dudeplayz) [inline .github/workflows/validation.yaml:175]:
```suggestion
      - name: Extract downloaded files
```

File: .github/workflows/validation.yaml [hunk 8]
```diff
@@ -61,25 +76,116 @@ jobs:
...
-          check_run_annotations: all tests, skipped tests
+      - name: Set Failure Status
+        if: ${{ always() && needs.changes.outputs.validation-required == 'true' && (needs.build-and-test.result != 'success' || needs.e2e-dev-tests.result != 'success' || needs.e2e-prod-tests.result != 'success') }}
+        run: |
+          echo "🚫 THERE ARE TEST MODULES WITH FAILURES or BEEN CANCELLED" | tee -a $GITHUB_STEP_SUMMARY
+          exit 1
+      - name: Set validation skipped status
+        if: ${{ always() && needs.changes.outputs.validation-required == 'false' }}
+        run: |
+          echo "✅ Pull request contains changes on files ignored by validation workflow" | tee -a $GITHUB_STEP_SUMMARY && exit 0
```
Comment (Dudeplayz) [inline .github/workflows/validation.yaml:191]:
```suggestion
          echo "✅ Pull request only contains changes on files ignored by validation workflow" | tee -a $GITHUB_STEP_SUMMARY && exit 0
```

**Review by Dudeplayz — CHANGES_REQUESTED** [review]

File: .github/workflows/validation.yaml (hunk 1 above)
Comment (mcollovati) [reply .github/workflows/validation.yaml:81]: Yes, it stands for STDIN

**Review by Dudeplayz — APPROVED** [review]

**Pull request merged**